README.rst
numinwords - Convert numbers to words in multiple languages
==========================================================

.. image:: https://img.shields.io/pypi/v/numinwords.svg
   :target: https://pypi.python.org/pypi/numinwords

.. image:: https://travis-ci.org/savoirfairelinux/numinwords.svg?branch=master
    :target: https://travis-ci.org/savoirfairelinux/numinwords

.. image:: https://coveralls.io/repos/github/savoirfairelinux/numinwords/badge.svg?branch=master
    :target: https://coveralls.io/github/savoirfairelinux/numinwords?branch=master


``numinwords`` is a library that converts numbers like ``42`` to words like ``forty-two``.
It supports multiple languages (see the list below for full list
of languages) and can even generate ordinal numbers like ``forty-second``
(although this last feature is a bit buggy for some languages at the moment).

The project is hosted on GitHub_. Contributions are welcome.

.. _GitHub: https://github.com/savoirfairelinux/numinwords

Installation
------------

The easiest way to install ``numinwords`` is to use pip::

    pip install numinwords

Otherwise, you can download the source package and then execute::

    python setup.py install

The test suite in this library is new, so it's rather thin, but it can be run with::

    python setup.py test

To run the full CI test suite which includes linting and multiple python environments::

    pip install tox
    tox

Usage
-----
Command line::

    $ numinwords 10001
    ten thousand and one
    $ numinwords 24,120.10
    twenty-four thousand, one hundred and twenty point one
    $ numinwords 24,120.10 -l es
    veinticuatro mil ciento veinte punto uno
    $numinwords 2.14 -l es --to currency
    dos euros con catorce centimos

In code there's only one function to use::

    >>> from numinwords import numinwords
    >>> numinwords(42)
    forty-two
    >>> numinwords(42, to='ordinal')
    forty-second
    >>> numinwords(42, lang='fr')
    quarante-deux

Besides the numerical argument, there are two main optional arguments.

**to:** The converter to use. Supported values are:

* ``cardinal`` (default)
* ``ordinal``
* ``ordinal_num``
* ``year``
* ``currency``

**lang:** The language in which to convert the number. Supported values are:

* ``en`` (English, default)
* ``ar`` (Arabic)
* ``cz`` (Czech)
* ``de`` (German)
* ``dk`` (Danish)
* ``en_GB`` (English - Great Britain)
* ``en_IN`` (English - India)
* ``es`` (Spanish)
* ``es_CO`` (Spanish - Colombia)
* ``es_VE`` (Spanish - Venezuela)
* ``eu`` (EURO)
* ``fi`` (Finnish)
* ``fr`` (French)
* ``fr_CH`` (French - Switzerland)
* ``fr_BE`` (French - Belgium)
* ``fr_DZ`` (French - Algeria)
* ``he`` (Hebrew)
* ``id`` (Indonesian)
* ``it`` (Italian)
* ``ja`` (Japanese)
* ``kn`` (Kannada)
* ``ko`` (Korean)
* ``kz`` (Kazakh)
* ``lt`` (Lithuanian)
* ``lv`` (Latvian)
* ``no`` (Norwegian)
* ``pl`` (Polish)
* ``pt`` (Portuguese)
* ``pt_BR`` (Portuguese - Brazilian)
* ``sl`` (Slovene)
* ``sr`` (Serbian)
* ``ro`` (Romanian)
* ``ru`` (Russian)
* ``te`` (Telugu)
* ``tr`` (Turkish)
* ``th`` (Thai)
* ``vi`` (Vietnamese)
* ``nl`` (Dutch)
* ``uk`` (Ukrainian)

You can supply values like ``fr_FR``; if the country doesn't exist but the
language does, the code will fall back to the base language (i.e. ``fr``). If
you supply an unsupported language, ``NotImplementedError`` is raised.
Therefore, if you want to call ``numinwords`` with a fallback, you can do::

    try:
        return numinwords(42, lang=mylang)
    except NotImplementedError:
        return numinwords(42, lang='en')

Additionally, some converters and languages support other optional arguments
that are needed to make the converter useful in practice.

Wiki
----
For additional information on some localization please check the Wiki_.
And feel free to propose wiki enhancement.

.. _Wiki: https://github.com/savoirfairelinux/numinwords/wiki

History
-------

``numinwords`` is based on an old library, ``pynum2word``, created by Taro Ogawa
in 2003. Unfortunately, the library stopped being maintained and the author
can't be reached. There was another developer, Marius Grigaitis, who in 2011
added Lithuanian support, but didn't take over maintenance of the project.

I am thus basing myself on Marius Grigaitis' improvements and re-publishing
``pynum2word`` as ``numinwords``.

Virgil Dupras, Savoir-faire Linux


CHANGES.rst
Changelog
=========


Version 0.5.10 -- 2019/05/12
----------------------------

* Add Kannada language localization (#243)
* Revert some copyrights changed by mistake (#254)
* Add indian rupee to the supported currencies (#248)
* Improve currency functions for German and French (#247)
* Improve Slovene localization (#246)
* Improve Spanish localization (#240)
* Fix typo 'seperator' on source code (#238)
* Convert string to decimal values (#223)
* Improve German localization and test coverage (#237)
* Improve Polish localization (#233)
* Fix ordinal number for French ending on 1 (#236)

Version 0.5.9 -- 2019/01/10
---------------------------

* Fix encoding issue on release 0.5.8 (#229)
* Improve Polish localization (#228)


Version 0.5.8 -- 2018/11/17
---------------------------

* Add Portuguese (Portugal) localization (#198)
* Add a command line tool to use numinwords
* Use language iso code for Vietnamese
* Improve Korean localization (#219)
* Improve Serbian (Latin) localization (#207)
* Improve testing setup (#220)
* Improve German localization (#214) (#222)
* Improve Romanian localization (#215)
* Improve Spanish localization (#187) (#200)
* Improve Russian localization (#211) (#212)
* Improve French localization (23902ab)
* Improve Arabic localization (#176)
* Improve Lithuanian and Latvian localization (#185)
* Improve Ukrainian localization (#183)


Version 0.5.7 -- 2018/06/27
---------------------------

* Add Finnish localization. (#170)
* Add Japanese localization. (#171)
* Add belgian-french localization. (#151)
* Add Czech localization. (#154) 
* Add Thai localization. (#139)
* Improve English localization. (#144) 
* Improve Spanish localization. (#167)
* Improve Italian localization. (#143)
* Improve documentation. (#155, #145, #174)

Version 0.5.6 -- 2017/11/22
---------------------------

* Refactor to_currency (#135)
* Allow the use of other convertes to_currency, to_year (#95)
* Fix code to respect PEP8 (#98, #105)
* Add Slovene localization (#97)
* Add Ukrainian localization (#93)
* Add Dutch localization (#91)
* Add Algeria-French localization (#86)
* Add Turkish localization (#85)

Version 0.5.5 -- 2017/07/02
---------------------------

* Add Arabic localization (#72)
* Add Spanish-Colombian and Spanish-Venezuelan localization (#67)
* Add VietNam localization (#61)
* Add Italian localization (#56, #59)
* Improve Russian localization (#62)
* Improve Polish localization (#58)

Version 0.5.4 -- 2016/10/18
---------------------------

* Tons of new languages!
* Add Polish localization. (#23)
* Add Swiss-French localization. (#38)
* Add Russian localization. (#28, #46, #48)
* Add Indonesian localization. (#29)
* Add Norwegian localization. (#33)
* Add Danish localization. (#40)
* Add Brazilian localization. (#37, #47)
* Improve German localization. (#25, #27, #49)
* Improve Lithuanian localization. (#52)
* Improve floating point spelling. (#24)

Version 0.5.3 -- 2015/06/09
---------------------------

* Fix packaging issues. (#21, #22)

Version 0.5.2 -- 2015/01/23
---------------------------

* Added Latvian localization. (#9)
* Improved Spanish localization. (#10, #13, #14)
* Improved Lithuanian localization. (#12)

Version 0.5.1 -- 2014/03/14
---------------------------

* Added Python 3 support with 2to3. (#3)
* Fixed big numbers in spanish. (#2)
* Fixed bugs in tanslation from 30 to 40 in spanish. (#4)
* Fixed word joining in english. (#8)

Version 0.5.0 -- 2013/05/28
---------------------------

* Created ``numinwords`` based on the old ``pynum2word`` project.


tests/test_ar.py
# -*- coding: utf-8 -*-
# Copyright (c) 2003, Taro Ogawa.  All Rights Reserved.
# Copyright (c) 2013, Savoir-faire Linux inc.  All Rights Reserved.

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301 USA

from unittest import TestCase

from numinwords import numinwords


class numinwordsARTest(TestCase):

    def test_default_currency(self):
        self.assertEqual(numinwords(1, to='currency', lang='ar'), 'واحد ريال')
        self.assertEqual(numinwords(2, to='currency', lang='ar'),
                         'اثنان ريالان')
        self.assertEqual(numinwords(10, to='currency', lang='ar'),
                         'عشرة ريالات')
        self.assertEqual(numinwords(
            100, to='currency', lang='ar'), 'مائة ريال')
        self.assertEqual(numinwords(652.12, to='currency', lang='ar'),
                         'ستمائة و اثنان و خمسون ريالاً و اثنتا عشرة هللة')
        self.assertEqual(numinwords(324, to='currency', lang='ar'),
                         'ثلاثمائة و أربعة و عشرون ريالاً')
        self.assertEqual(numinwords(2000, to='currency', lang='ar'),
                         'ألفا ريال')
        self.assertEqual(numinwords(541, to='currency', lang='ar'),
                         'خمسمائة و واحد و أربعون ريالاً')
        self.assertEqual(numinwords(10000, to='currency', lang='ar'),
                         'عشرة آلاف ريال')
        self.assertEqual(numinwords(20000.12, to='currency', lang='ar'),
                         'عشرون ألف ريال و اثنتا عشرة هللة')
        self.assertEqual(numinwords(1000000, to='currency', lang='ar'),
                         'واحد مليون ريال')
        val = 'تسعمائة و ثلاثة و عشرون ألفاً  و أربعمائة و أحد عشر ريالاً'
        self.assertEqual(numinwords(923411, to='currency', lang='ar'), val)
        self.assertEqual(numinwords(63411, to='currency', lang='ar'),
                         'ثلاثة و ستون ألفاً  و أربعمائة و أحد عشر ريالاً')
        self.assertEqual(numinwords(1000000.99, to='currency', lang='ar'),
                         'واحد مليون ريال و تسع و تسعون هللة')

    def test_currency_parm(self):
        self.assertEqual(
            numinwords(1, to='currency', lang='ar', currency="KWD"),
            'واحد دينار')
        self.assertEqual(
            numinwords(10, to='currency', lang='ar', currency="EGP"),
            'عشرة جنيهات')
        self.assertEqual(
            numinwords(20000.12, to='currency', lang='ar', currency="EGP"),
            'عشرون ألف جنيه و اثنتا عشرة قرش')
        self.assertEqual(
            numinwords(923411, to='currency', lang='ar', currency="SR"),
            'تسعمائة و ثلاثة و عشرون ألفاً  و أربعمائة و أحد عشر ريالاً')
        self.assertEqual(
            numinwords(1000000.99, to='currency', lang='ar', currency="KWD"),
            'واحد مليون دينار و تسع و تسعون فلس')

    def test_ordinal(self):
        self.assertEqual(numinwords(1, to='ordinal', lang='ar'), 'اول')
        self.assertEqual(numinwords(2, to='ordinal', lang='ar'), 'ثاني')
        self.assertEqual(numinwords(3, to='ordinal', lang='ar'), 'ثالث')
        self.assertEqual(numinwords(4, to='ordinal', lang='ar'), 'رابع')
        self.assertEqual(numinwords(5, to='ordinal', lang='ar'), 'خامس')
        self.assertEqual(numinwords(6, to='ordinal', lang='ar'), 'سادس')
        self.assertEqual(numinwords(9, to='ordinal', lang='ar'), 'تاسع')
        self.assertEqual(numinwords(20, to='ordinal', lang='ar'), 'عشرون')
        self.assertEqual(numinwords(94, to='ordinal', lang='ar'),
                         'أربع و تسعون')
        self.assertEqual(numinwords(102, to='ordinal', lang='ar'),
                         'مائة و اثنان')
        self.assertEqual(
            numinwords(923411, to='ordinal_num', lang='ar'),
            'تسعمائة و ثلاثة و عشرون ألفاً  و أربعمائة و أحد عشر')

    def test_cardinal(self):
        self.assertEqual(numinwords(12, to='cardinal', lang='ar'), 'اثنا عشر')
        self.assertEqual(numinwords(-8324, to='cardinal', lang='ar'),
                         'سالب ثمانية آلاف  و ثلاثمائة و أربعة و عشرون')
        self.assertEqual(
            numinwords(3431.12, to='cardinal', lang='ar'),
            'ثلاثة آلاف  و أربعمائة و واحد و ثلاثون  , اثنتا عشرة')
        self.assertEqual(numinwords(431, to='cardinal', lang='ar'),
                         'أربعمائة و واحد و ثلاثون')
        self.assertEqual(numinwords(94231, to='cardinal', lang='ar'),
                         'أربعة و تسعون ألفاً  و مئتان و واحد و ثلاثون')
        self.assertEqual(numinwords(1431, to='cardinal', lang='ar'),
                         'واحد ألف  و أربعمائة و واحد و ثلاثون')

    def test_prefix_and_suffix(self):
        self.assertEqual(numinwords(645, to='currency',
                                    lang='ar', prefix="فقط", suffix="لاغير"),
                         'فقط ستمائة و خمسة و أربعون ريالاً لاغير')

    def test_year(self):
        self.assertEqual(numinwords(2000, to='year', lang='ar'), 'ألفا')

    def test_max_numbers(self):
        with self.assertRaises(Exception) as context:
            numinwords(10 ** 36, to='year', lang='ar')

        self.assertTrue('Too large' in str(context.exception))


tests/test_cz.py
# -*- coding: utf-8 -*-
# Copyright (c) 2003, Taro Ogawa.  All Rights Reserved.
# Copyright (c) 2013, Savoir-faire Linux inc.  All Rights Reserved.

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301 USA

from __future__ import unicode_literals

from unittest import TestCase

from numinwords import numinwords


class numinwordsCZTest(TestCase):
    def test_cardinal(self):
        self.assertEqual(numinwords(100, lang='cz'), "sto")
        self.assertEqual(numinwords(101, lang='cz'), "sto jedna")
        self.assertEqual(numinwords(110, lang='cz'), "sto deset")
        self.assertEqual(numinwords(115, lang='cz'), "sto patnáct")
        self.assertEqual(numinwords(123, lang='cz'), "sto dvacet tři")
        self.assertEqual(numinwords(1000, lang='cz'), "tisíc")
        self.assertEqual(numinwords(1001, lang='cz'), "tisíc jedna")
        self.assertEqual(numinwords(2012, lang='cz'), "dva tisíce dvanáct")
        self.assertEqual(
            numinwords(12519.85, lang='cz'),
            "dvanáct tisíc pětset devatenáct celá osmdesát pět"
        )
        self.assertEqual(
            numinwords(123.50, lang='cz'),
            "sto dvacet tři celá pět"
        )
        self.assertEqual(
            numinwords(1234567890, lang='cz'),
            "miliarda dvěstě třicet čtyři miliony pětset šedesát "
            "sedm tisíc osmset devadesát"
        )
        self.assertEqual(
            numinwords(215461407892039002157189883901676, lang='cz'),
            "dvěstě patnáct quintillionů čtyřista šedesát jedna kvadriliard "
            "čtyřista sedm kvadrilionů osmset devadesát dva triliardy třicet "
            "devět trilionů dva biliardy sto padesát sedm bilionů sto "
            "osmdesát devět miliard osmset osmdesát tři miliony "
            "devětset jedna tisíc šestset sedmdesát šest"
        )
        self.assertEqual(
            numinwords(719094234693663034822824384220291, lang='cz'),
            "sedmset devatenáct quintillionů devadesát "
            "čtyři kvadriliardy dvěstě třicet čtyři "
            "kvadriliony šestset devadesát tři triliardy "
            "šestset šedesát tři triliony třicet čtyři biliardy osmset "
            "dvacet dva biliony osmset dvacet čtyři "
            "miliardy třista osmdesát čtyři miliony dvěstě dvacet "
            "tisíc dvěstě devadesát jedna"
        )

    def test_to_ordinal(self):
        # @TODO: implement to_ordinal
        with self.assertRaises(NotImplementedError):
            numinwords(1, lang='cz', to='ordinal')

    def test_currency(self):
        self.assertEqual(
            numinwords(10.0, lang='cz', to='currency', currency='EUR'),
            "deset euro, nula centů")
        self.assertEqual(
            numinwords(1.0, lang='cz', to='currency', currency='CZK'),
            "jedna koruna, nula haléřů")
        self.assertEqual(
            numinwords(1234.56, lang='cz', to='currency', currency='EUR'),
            "tisíc dvěstě třicet čtyři euro, padesát šest centů")
        self.assertEqual(
            numinwords(1234.56, lang='cz', to='currency', currency='CZK'),
            "tisíc dvěstě třicet čtyři koruny, padesát šest haléřů")
        self.assertEqual(
            numinwords(101.11, lang='cz', to='currency', currency='EUR',
                       separator=' a'),
            "sto jedna euro a jedenáct centů")
        self.assertEqual(
            numinwords(101.21, lang='cz', to='currency', currency='CZK',
                       separator=' a'),
            "sto jedna korun a dvacet jedna haléřů"
        )
        self.assertEqual(
            numinwords(-12519.85, lang='cz', to='currency', cents=False),
            "mínus dvanáct tisíc pětset devatenáct euro, 85 centů"
        )
        self.assertEqual(
            numinwords(123.50, lang='cz', to='currency', currency='CZK',
                       separator=' a'),
            "sto dvacet tři koruny a padesát haléřů"
        )
        self.assertEqual(
            numinwords(19.50, lang='cz', to='currency', cents=False),
            "devatenáct euro, 50 centů"
        )


tests/test_de.py
# -*- coding: utf-8 -*-
# Copyright (c) 2003, Taro Ogawa.  All Rights Reserved.
# Copyright (c) 2013, Savoir-faire Linux inc.  All Rights Reserved.

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301 USA

from __future__ import unicode_literals

from unittest import TestCase

from numinwords import numinwords

TEST_CASES_TO_CURRENCY_EUR = (
    (1.00, 'ein Euro und null Cent'),
    (2.01, 'zwei Euro und ein Cent'),
    (8.10, 'acht Euro und zehn Cent'),
    (12.26, 'zwölf Euro und sechsundzwanzig Cent'),
    (21.29, 'einundzwanzig Euro und neunundzwanzig Cent'),
    (81.25, 'einundachtzig Euro und fünfundzwanzig Cent'),
    (100.00, 'einhundert Euro und null Cent'),
)

TEST_CASES_TO_CURRENCY_USD = (
    (1.00, 'ein Dollar und null Cent'),
    (2.01, 'zwei Dollar und ein Cent'),
    (8.10, 'acht Dollar und zehn Cent'),
    (12.26, 'zwölf Dollar und sechsundzwanzig Cent'),
    (21.29, 'einundzwanzig Dollar und neunundzwanzig Cent'),
    (81.25, 'einundachtzig Dollar und fünfundzwanzig Cent'),
    (100.00, 'einhundert Dollar und null Cent'),
)

TEST_CASES_TO_CURRENCY_GBP = (
    (1.00, 'ein Pfund und null Pence'),
    (2.01, 'zwei Pfund und ein Penny'),
    (8.10, 'acht Pfund und zehn Pence'),
    (12.26, 'zwölf Pfund und sechsundzwanzig Pence'),
    (21.29, 'einundzwanzig Pfund und neunundzwanzig Pence'),
    (81.25, 'einundachtzig Pfund und fünfundzwanzig Pence'),
    (100.00, 'einhundert Pfund und null Pence'),
)

TEST_CASES_TO_CURRENCY_DEM = (
    (1.00, 'ein Mark und null Pfennig'),
    (2.01, 'zwei Mark und ein Pfennig'),
    (8.10, 'acht Mark und zehn Pfennig'),
    (12.26, 'zwölf Mark und sechsundzwanzig Pfennig'),
    (21.29, 'einundzwanzig Mark und neunundzwanzig Pfennig'),
    (81.25, 'einundachtzig Mark und fünfundzwanzig Pfennig'),
    (100.00, 'einhundert Mark und null Pfennig'),
)


class numinwordsDETest(TestCase):

    def test_ordinal_less_than_twenty(self):
        self.assertEqual(numinwords(0, ordinal=True, lang='de'), "nullte")
        self.assertEqual(numinwords(1, ordinal=True, lang='de'), "erste")
        self.assertEqual(numinwords(7, ordinal=True, lang='de'), "siebte")
        self.assertEqual(numinwords(8, ordinal=True, lang='de'), "achte")
        self.assertEqual(numinwords(12, ordinal=True, lang='de'), "zwölfte")
        self.assertEqual(numinwords(17, ordinal=True, lang='de'), "siebzehnte")

    def test_ordinal_more_than_twenty(self):
        self.assertEqual(
            numinwords(81, ordinal=True, lang='de'), "einundachtzigste"
        )

    def test_ordinal_at_crucial_number(self):
        self.assertEqual(
            numinwords(100, ordinal=True, lang='de'), "hundertste"
        )
        self.assertEqual(
            numinwords(1000, ordinal=True, lang='de'), "tausendste"
        )
        self.assertEqual(
            numinwords(4000, ordinal=True, lang='de'), "viertausendste"
        )
        self.assertEqual(
            numinwords(1000000, ordinal=True, lang='de'), "millionste"
        )
        self.assertEqual(
            numinwords(2000000, ordinal=True, lang='de'), "zweimillionste"
        )
        self.assertEqual(
            numinwords(1000000000, ordinal=True, lang='de'), "milliardste"
        )
        self.assertEqual(
            numinwords(5000000000, ordinal=True, lang='de'),
            "fünfmilliardste"
        )

    def test_cardinal_at_some_numbers(self):
        self.assertEqual(numinwords(100, lang='de'), "einhundert")
        self.assertEqual(numinwords(1000, lang='de'), "eintausend")
        self.assertEqual(numinwords(5000, lang='de'), "fünftausend")
        self.assertEqual(numinwords(10000, lang='de'), "zehntausend")
        self.assertEqual(numinwords(1000000, lang='de'), "eine Million")
        self.assertEqual(numinwords(2000000, lang='de'), "zwei Millionen")
        self.assertEqual(numinwords(4000000000, lang='de'), "vier Milliarden")
        self.assertEqual(numinwords(1000000000, lang='de'), "eine Milliarde")

    def test_cardinal_for_decimal_number(self):
        self.assertEqual(
            numinwords(3.486, lang='de'), "drei Komma vier acht sechs"
        )

    def test_giant_cardinal_for_merge(self):
        self.assertEqual(
            numinwords(4500072900000111, lang='de'),
            "vier Billiarden fünfhundert Billionen " +
            "zweiundsiebzig Milliarden neunhundert Millionen einhundertelf"
        )

    def test_ordinal_num(self):
        self.assertEqual(numinwords(7, to="ordinal_num", lang='de'), "7.")
        self.assertEqual(numinwords(81, to="ordinal_num", lang='de'), "81.")

    def test_ordinal_for_negative_numbers(self):
        self.assertRaises(TypeError, numinwords, -12, ordinal=True, lang='de')

    def test_ordinal_for_floating_numbers(self):
        self.assertRaises(TypeError, numinwords, 2.453, ordinal=True,
                          lang='de')

    def test_currency_eur(self):
        for test in TEST_CASES_TO_CURRENCY_EUR:
            self.assertEqual(
                numinwords(test[0], lang='de', to='currency', currency='EUR'),
                test[1]
            )

    def test_currency_usd(self):
        for test in TEST_CASES_TO_CURRENCY_USD:
            self.assertEqual(
                numinwords(test[0], lang='de', to='currency', currency='USD'),
                test[1]
            )

    def test_currency_dem(self):
        for test in TEST_CASES_TO_CURRENCY_DEM:
            self.assertEqual(
                numinwords(test[0], lang='de', to='currency', currency='DEM'),
                test[1]
            )

    def test_currency_gbp(self):
        for test in TEST_CASES_TO_CURRENCY_GBP:
            self.assertEqual(
                numinwords(test[0], lang='de', to='currency', currency='GBP'),
                test[1]
            )

    def test_year(self):
        self.assertEqual(numinwords(2002, to='year', lang='de'),
                         'zweitausendzwei')

    def test_year_before_2000(self):
        self.assertEqual(numinwords(1780, to='year', lang='de'),
                         'siebzehnhundertachtzig')


numinwords/lang_TH.py
# -*- coding: utf-8 -*-
# Copyright (c) 2003, Taro Ogawa.  All Rights Reserved.
# Copyright (c) 2013, Savoir-faire Linux inc.  All Rights Reserved.

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301 USA

from __future__ import unicode_literals

from numinwords.base import Num2Word_Base
from numinwords.currency import parse_currency_parts
from numinwords.utils import splitbyx


class Num2Word_TH(Num2Word_Base):

    def setup(self):
        self.negword = 'ติดลบ'
        self.pointword = 'จุด'

        self.CURRENCY_FORMS = {
            'THB': (('บาท', 'บาท'), ('สตางค์', 'สตางค์')),
            'USD': (('ดอลลาร์', 'ดอลลาร์'), ('เซนต์', 'เซนต์')),
            'EUR': (('ยูโร', 'ยูโร'), ('เซนต์', 'เซนต์')),
        }

        self.high_numwords = []

        self.mid_numwords = ['', 'สิบ', 'ร้อย', 'พัน', 'หมื่น', 'แสน', 'ล้าน']

        self.low_numwords = [
            'ศูนย์', 'หนึ่ง', 'สอง', 'สาม', 'สี่',
            'ห้า', 'หก', 'เจ็ด', 'แปด', 'เก้า'
        ]

    def set_high_numwords(self, high_numwords):
        pass

    def set_mid_numwords(self, mid_numwords):
        pass

    def splitnum(self, six_num):
        length = len(six_num) > 1
        word_num = ''

        for index, num in enumerate(map(int, six_num)):
            if num:
                if index:
                    word_num = self.mid_numwords[index] + word_num

                if length and num == 1 and index == 0:
                    word_num += 'เอ็ด'
                elif index == 1 and num == 2:
                    word_num = 'ยี่' + word_num
                elif index != 1 or num != 1:
                    word_num = self.low_numwords[num] + word_num

            elif num == 0 and index == 0 and length == 0:
                word_num = self.low_numwords[0]

        return word_num

    def split_six(self, num_txt):
        result = splitbyx(num_txt, 6, format_int=False)
        result = list(result)[::-1]
        number_list = []
        for i in result:
            number_list.append(i[::-1])
        return number_list

    def add_text_million(self, word_num):
        result = ''

        for index, t in enumerate(reversed(word_num)):
            if index == 0:
                result = t
            else:
                result = result + 'ล้าน' + t

        return result

    def round_2_decimal(self, number):
        integer, cents, negative = parse_currency_parts(
            number, is_int_with_cents=False
        )
        integer = '{}'.format(integer)
        cents = '{}'.format(cents)

        if len(cents) < 2:
            add_zero = 2 - len(cents)
            cents = ('0' * add_zero) + cents

        text_num = integer + '.' + cents

        return text_num, negative

    def left_num_to_text(self, number):

        left_num_list = self.split_six(number)

        left_text_list = []
        for i in left_num_list:
            left_text_list.append(self.splitnum(i))

        left_text = self.add_text_million(left_text_list)
        return left_text

    def to_cardinal(self, number):
        negative = number < 0

        pre, post = self.float2tuple(number)
        precision = self.precision
        pre = '{}'.format(pre)
        post = '{}'.format(post)

        if negative:
            pre = pre.lstrip('-')

        if len(post) < precision:
            add_zero = precision - len(post)
            post = ('0' * add_zero) + post

        result = self.left_num_to_text(pre)

        right_text = ''
        if not post == '0':
            for i in map(int, post):
                right_text = right_text + self.low_numwords[i]
            result = result + 'จุด' + right_text

        if negative:
            result = 'ติดลบ' + result

        return result

    def to_ordinal(self, number):
        return self.to_cardinal(number)

    def to_currency(self, number, currency='THB'):

        number, negative = self.round_2_decimal(number)

        split_num = number.split('.')

        left_num = split_num[0]
        left_text = self.left_num_to_text(left_num)

        right_num = split_num[1]
        right_text = self.splitnum(right_num[::-1].rstrip('0'))

        try:
            cr1, cr2 = self.CURRENCY_FORMS[currency]

        except KeyError:
            raise NotImplementedError(
                'Currency code "%s" not implemented for "%s"' %
                (currency, self.__class__.__name__))

        if right_num == '00':
            if currency == 'THB':
                result = left_text + cr1[0] + 'ถ้วน'
            else:
                result = left_text + cr1[0]
        else:
            if left_num == '0':
                result = right_text + cr2[0]
            else:
                result = left_text + cr1[0] + right_text + cr2[0]

        if negative:
            result = self.negword + result

        return result


numinwords/__init__.py
# -*- coding: utf-8 -*-
# Copyright (c) 2003, Taro Ogawa.  All Rights Reserved.
# Copyright (c) 2013, Savoir-faire Linux inc.  All Rights Reserved.

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301 USA

from __future__ import unicode_literals

from . import (lang_AR, lang_CZ, lang_DE, lang_DK, lang_EN, lang_EN_IN,
               lang_ES, lang_ES_CO, lang_ES_NI, lang_ES_VE, lang_FI, lang_FR,
               lang_FR_BE, lang_FR_CH, lang_FR_DZ, lang_HE, lang_ID, lang_IT,
               lang_JA, lang_KN, lang_KO, lang_KZ, lang_LT, lang_LV, lang_NL,
               lang_NO, lang_PL, lang_PT, lang_PT_BR, lang_RO, lang_RU,
               lang_SL, lang_SR, lang_TE, lang_TH, lang_TR, lang_UK, lang_VI)

CONVERTER_CLASSES = {
    'ar': lang_AR.Num2Word_AR(),
    'cz': lang_CZ.Num2Word_CZ(),
    'en': lang_EN.Num2Word_EN(),
    'en_IN': lang_EN_IN.Num2Word_EN_IN(),
    'fr': lang_FR.Num2Word_FR(),
    'fr_CH': lang_FR_CH.Num2Word_FR_CH(),
    'fr_BE': lang_FR_BE.Num2Word_FR_BE(),
    'fr_DZ': lang_FR_DZ.Num2Word_FR_DZ(),
    'de': lang_DE.Num2Word_DE(),
    'fi': lang_FI.Num2Word_FI(),
    'es': lang_ES.Num2Word_ES(),
    'es_CO': lang_ES_CO.Num2Word_ES_CO(),
    'es_NI': lang_ES_NI.Num2Word_ES_NI(),
    'es_VE': lang_ES_VE.Num2Word_ES_VE(),
    'id': lang_ID.Num2Word_ID(),
    'ja': lang_JA.Num2Word_JA(),
    'kn': lang_KN.Num2Word_KN(),
    'ko': lang_KO.Num2Word_KO(),
    'kz': lang_KZ.Num2Word_KZ(),
    'lt': lang_LT.Num2Word_LT(),
    'lv': lang_LV.Num2Word_LV(),
    'pl': lang_PL.Num2Word_PL(),
    'ro': lang_RO.Num2Word_RO(),
    'ru': lang_RU.Num2Word_RU(),
    'sl': lang_SL.Num2Word_SL(),
    'sr': lang_SR.Num2Word_SR(),
    'no': lang_NO.Num2Word_NO(),
    'dk': lang_DK.Num2Word_DK(),
    'pt': lang_PT.Num2Word_PT(),
    'pt_BR': lang_PT_BR.Num2Word_PT_BR(),
    'he': lang_HE.Num2Word_HE(),
    'it': lang_IT.Num2Word_IT(),
    'vi': lang_VI.Num2Word_VI(),
    'th': lang_TH.Num2Word_TH(),
    'tr': lang_TR.Num2Word_TR(),
    'nl': lang_NL.Num2Word_NL(),
    'uk': lang_UK.Num2Word_UK(),
    'te': lang_TE.Num2Word_TE()
}


CONVERTES_TYPES = ['cardinal', 'ordinal', 'ordinal_num', 'year', 'currency']


def numinwords(number, ordinal=False, lang='en', to='cardinal', **kwargs):
    # We try the full language first
    if lang not in CONVERTER_CLASSES:
        # ... and then try only the first 2 letters
        lang = lang[:2]
    if lang not in CONVERTER_CLASSES:
        raise NotImplementedError()
    converter = CONVERTER_CLASSES[lang]

    if isinstance(number, str):
        number = converter.str_to_number(number)

    # backwards compatible
    if ordinal:
        return converter.to_ordinal(number)

    if to not in CONVERTES_TYPES:
        raise NotImplementedError()

    return getattr(converter, 'to_{}'.format(to))(number, **kwargs)
